apis/s3/src/main/java/org/jclouds/s3/filters/RequestAuthorizeSignatureV4.java

```diff
@@ -0,0 +1,461 @@
+     * joining them, in order, separating key value pairs with an '&'.
+     *
+     * @param queryString The query string parameters to be canonicalized.
+     * @return A canonicalized form for the specified query string parameters.
+     */
+    protected String getCanonicalizedQueryString(String queryString) {
+        Multimap<String, String> params = queryParser().apply(queryString);
+        SortedMap<String, String> sorted = new TreeMap<String, String>();
+        if (params == null) {
+            return "";
+        }
+        Iterator<Map.Entry<String, String>> pairs = params.entries().iterator();
+        while (pairs.hasNext()) {
+            Map.Entry<String, String> pair = pairs.next();
+            String key = pair.getKey();
+            String value = pair.getValue();
+            sorted.put(urlEncode(key), urlEncode(value));
+        }
+
+        StringBuilder builder = new StringBuilder();
+        pairs = sorted.entrySet().iterator();
+        while (pairs.hasNext()) {
+            Map.Entry<String, String> pair = pairs.next();
+            builder.append(pair.getKey());
+            builder.append("=");
+            builder.append(pair.getValue());
+            if (pairs.hasNext()) {
+                builder.append("&");
+            }
+        }
+
+        return builder.toString();
+    }
+
+    private static final TimeZone GMT = TimeZone.getTimeZone("GMT");
+    private static final DateFormat timestampFormat;
+    private static final DateFormat dateFormat;
+
+    static {
+        timestampFormat = new SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'");
+        dateFormat = new SimpleDateFormat("yyyyMMdd");
+        timestampFormat.setTimeZone(GMT);
+        dateFormat.setTimeZone(GMT);
+    }
+
+
+    /**
+     * Encode a string for use in the path of a URL; uses URLEncoder.encode,
+     * (which encodes a string for use in the query portion of a URL), then
+     * applies some postfilters to fix things up per the RFC. Can optionally
+     * handle strings which are meant to encode a path (ie include '/'es
+     * which should NOT be escaped).
+     *
+     * @param value the value to encode
+     * @return the encoded value
+     */
+    public static String urlEncode(final String value) {
+        if (value == null) {
+            return "";
+        }
+
+        try {
+            String encoded = URLEncoder.encode(value, DEFAULT_ENCODING);
+
+            Matcher matcher = ENCODED_CHARACTERS_PATTERN.matcher(encoded);
+            StringBuffer buffer = new StringBuffer(encoded.length());
+
+            while (matcher.find()) {
+                String replacement = matcher.group(0);
+
+                if ("+".equals(replacement)) {
+                    replacement = "%20";
+                } else if ("*".equals(replacement)) {
+                    replacement = "%2A";
+                } else if ("%7E".equals(replacement)) {
+                    replacement = "~";
+                }
```

Comment:
UrlEscapers.urlFormParameterEscaper isn's same as this urlEncode.
* URI encode every byte except the unreserved characters: 'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', and '~'.
* The space character is a reserved character and must be encoded as "%20" (and not as "+").
* Each Uri-encoded byte is formed by a '%' and the two-digit hexadecimal value of the byte.
* Letters in the hexadecimal value must be uppercase, for example "%1A".
* Encode the forward slash character, '/', everywhere except in the object key name. For example, if the object key name is photos/Jan/sample.jpg, the forward slash in the key name is not encoded.